src/utils/compatibility.js

```diff
@@ -0,0 +1,64 @@
+import mime from 'mime-types';
+
+export function resGetHeader(res, header) {
+  // for express
+  if (res.get) {
+    return res.get(header);
+  }
+  // for fastify
+  return res.getHeader(header);
+}
+
+export function resSetHeader(res, header, value) {
+  // for express
+  if (res.set) {
+    return res.set(header, value);
+  }
+  // for fastify
+  return res.setHeader(header, value);
+}
+
+// helper to get Content-Length for fastify
+export function getContentLength(content) {
+  let len = 0;
+  if (typeof content !== 'undefined') {
+    if (Buffer.isBuffer(content)) {
+      // get length of Buffer
+      len = content.length;
+    } else if (content.length < 1000) {
+      // just calculate length
+      len = Buffer.byteLength(content);
+    } else {
+      // convert content to Buffer and calculate
+      // eslint-disable-next-line no-param-reassign
+      content = Buffer.from(content);
+      len = content.length;
+    }
+  }
+  return len;
+}
+
+// helper to get Content-Type for fastify
+export function getUnknownContentType(content) {
+  return Buffer.isBuffer(content)
+    ? mime.contentType('bin')
+    : mime.contentType('html');
+}
+
+export function sendContent(res, content) {
+  // for express
+  if (res.send) {
+    return res.send(content);
+  }
+
+  // for fastify
+  const contentLength = getContentLength(content);
+  resSetHeader(res, 'Content-Length', contentLength);
+
+  if (!resGetHeader(res, 'Content-Type')) {
+    const unknownContentType = getUnknownContentType(content);
+    resSetHeader(res, 'Content-Type', unknownContentType);
+  }
+
+  return res.end(content);
+}
```

Comment:
We really need `sendContent` https://www.fastify.io/docs/v1.14.x/Reply/#senddata?

Reply:
@evilebottnawi Yep we need it: https://github.com/fastify/fastify/issues/319

Reply:
I am afraid it is not safe to do this https://github.com/fastify/fastify/pull/328/files#diff-f5fc3aad6335ea7ddc5204ffa960e59eR34

Reply:
@evilebottnawi `If you need to create a middleware you have to use the Node `req` and `res` instances` that's exactly what we do we use `.end` and `.setHeader` that are associated with Node `res`.

Reply:
Look at example of middleware https://github.com/fastify/fastify-static/blob/master/index.js#L93, I still don't undestand why we send `Content-Type` here, can you provide more information?

Reply:
Looks `reply.send` is available here

Reply:
we send `content-type` to pass the tests as express set the `content-type` by default, whenever I use `res.send` alone it gives `send is not a function` error.

Reply:
@EslamHiko so `fastify` do not send `content-type` by default? It is strange because in other middleware I don't find code for `content-type`

Reply:
@evilebottnawi oh sorry `fastify` don't send `content-length` header but it sends `Content-Type: text/html; charset=utf-8` for unknown content-type files instead of `Content-Type: application/octet-stream` like express.

Reply:
hm, what is Node.js (using built-in `http`/`https` module) default behavior, can you investigate?

Reply:
@evilebottnawi I just made a very simple Node.js server and it didn't provide the content type for unknown mime type files.
`server.js`
```js
const http = require('http');

const hostname = '127.0.0.1';
const port = 3000;
const fs = require('fs');
const file = fs.readFileSync('./file');
const server = http.createServer((req, res) => {
  res.end(file);
});

server.listen(port, hostname, () => {
  console.log(`Server running at http://${hostname}:${port}/`);
});
```
`file`
```
this is supposed to be unknown mime type file
```
![headers](https://user-images.githubusercontent.com/18349358/92325859-22c9fe80-f04e-11ea-83dd-17420d3a6103.PNG)

We can see here what express do for content types here : https://github.com/expressjs/express/blob/0a48e18056865364b2461b2ece7ccb2d1075d3c9/lib/response.js#L141
and that's what fastify do : 
https://github.com/fastify/fastify/blob/b9fcf3fae0317a54f4f32c5121d5099d1a4d0ed9/lib/reply.js#L131

Reply:
Some notes : 
- If I remove the content length setting section it fails in this test "should return the "200" code for the "HEAD" request to the bundle file"
```js
it('should return the "200" code for the "HEAD" request to the bundle file', (done) => {
          request(server)
            .head('/bundle.js')
            .expect(
              'Content-Length',
              instance.context.outputFileSystem
                .readFileSync(path.resolve(outputPath, 'bundle.js'))
                .byteLength.toString()
            )
            .expect('Content-Type', 'application/javascript; charset=utf-8')
            // eslint-disable-next-line no-undefined
            .expect(200, undefined, done);
        });
```
![Screenshot from 2020-09-06 14-56-27](https://user-images.githubusercontent.com/18349358/92326264-2d39c780-f051-11ea-8f7c-3291fdec4ed8.png)

- I get this when I comment the unknown content-type setting section.
```
 middleware (fastify) › basic › should work › should return "200" code code for the "GET" request to the file without extension

    expected "Content-Type" header field

      at Test._assertHeader (node_modules/supertest/lib/test.js:237:45)
      ...

  ● middleware (fastify) › basic › should work with difference requests › should work handle requests › should return the "200" code for the "GET" request for the "filename-name-with-dots/mono-v6.x.x" url

    expected "Content-Type" header field

      at Test._assertHeader (node_modules/supertest/lib/test.js:237:45)
      ...

  ● middleware (fastify) › basic › should work with difference requests › should work handle requests › should return the "200" code for the "GET" request for the "noextension" url

    expected "Content-Type" header field

      at Test._assertHeader (node_modules/supertest/lib/test.js:237:45)
      ...

  ● middleware (fastify) › basic › should work with difference requests › should work handle requests › should return the "200" code for the "GET" request for the "filename-name-with-dots/mono-v6.x.x" url

    expected "Content-Type" header field

      at Test._assertHeader (node_modules/supertest/lib/test.js:237:45)
      ...

  ● middleware (fastify) › basic › should work with difference requests › should work handle requests › should return the "200" code for the "GET" request for the "noextension" url

    expected "Content-Type" header field

      at Test._assertHeader (node_modules/supertest/lib/test.js:237:45)
      ...

  ● middleware (fastify) › basic › should work with difference requests › should work handle requests › should return the "200" code for the "GET" request for the "filename-name-with-dots/mono-v6.x.x" url

    expected "Content-Type" header field

      at Test._assertHeader (node_modules/supertest/lib/test.js:237:45)
      ...

  ● middleware (fastify) › basic › should work with difference requests › should work handle requests › should return the "200" code for the "GET" request for the "noextension" url

    expected "Content-Type" header field

      at Test._assertHeader (node_modules/supertest/lib/test.js:237:45)
      ...

  ● middleware (fastify) › basic › should work with difference requests › should work handle requests › should return the "200" code for the "GET" request for the "filename-name-with-dots/mono-v6.x.x" url

    expected "Content-Type" header field

      at Test._assertHeader (node_modules/supertest/lib/test.js:237:45)
      ...

  ● middleware (fastify) › basic › should work with difference requests › should work handle requests › should return the "200" code for the "GET" request for the "noextension" url

    expected "Content-Type" header field

      at Test._assertHeader (node_modules/supertest/lib/test.js:237:45)
      ...

  ● middleware (fastify) › basic › should work with difference requests › should work handle requests › should return the "200" code for the "GET" request for the "filename-name-with-dots/mono-v6.x.x" url

    expected "Content-Type" header field

      at Test._assertHeader (node_modules/supertest/lib/test.js:237:45)
      ...

  ● middleware (fastify) › basic › should work with difference requests › should work handle requests › should return the "200" code for the "GET" request for the "noextension" url

    expected "Content-Type" header field

      at Test._assertHeader (node_modules/supertest/lib/test.js:237:45)
      ...

  ● middleware (fastify) › basic › should work with difference requests › should work handle requests › should return the "200" code for the "GET" request for the "filename-name-with-dots/mono-v6.x.x" url

    expected "Content-Type" header field

      at Test._assertHeader (node_modules/supertest/lib/test.js:237:45)
      ...

  ● middleware (fastify) › basic › should work with difference requests › should work handle requests › should return the "200" code for the "GET" request for the "noextension" url

    expected "Content-Type" header field

      at Test._assertHeader (node_modules/supertest/lib/test.js:237:45)
      ...

  ● middleware (fastify) › basic › should work with difference requests › should work handle requests › should return the "200" code for the "GET" request for the "filename-name-with-dots/mono-v6.x.x" url

    expected "Content-Type" header field

      at Test._assertHeader (node_modules/supertest/lib/test.js:237:45)
      ...

  ● middleware (fastify) › basic › should work with difference requests › should work handle requests › should return the "200" code for the "GET" request for the "noextension" url

    expected "Content-Type" header field

      at Test._assertHeader (node_modules/supertest/lib/test.js:237:45)
      ...

  ● middleware (fastify) › basic › should work with difference requests › should work handle requests › should return the "200" code for the "GET" request for the "filename-name-with-dots/mono-v6.x.x" url

    expected "Content-Type" header field

      at Test._assertHeader (node_modules/supertest/lib/test.js:237:45)
      ...

  ● middleware (fastify) › basic › should work with difference requests › should work handle requests › should return the "200" code for the "GET" request for the "noextension" url

    expected "Content-Type" header field

      at Test._assertHeader (node_modules/supertest/lib/test.js:237:45)
      ...

  ● middleware (fastify) › basic › should work with difference requests › should work handle requests › should return the "200" code for the "GET" request for the "filename-name-with-dots/mono-v6.x.x" url

    expected "Content-Type" header field

      at Test._assertHeader (node_modules/supertest/lib/test.js:237:45)
      ...

  ● middleware (fastify) › basic › should work with difference requests › should work handle requests › should return the "200" code for the "GET" request for the "noextension" url

    expected "Content-Type" header field

      at Test._assertHeader (node_modules/supertest/lib/test.js:237:45)
      ...

  ● middleware (fastify) › basic › should work with difference requests › should work handle requests › should return the "200" code for the "GET" request for the "filename-name-with-dots/mono-v6.x.x" url

    expected "Content-Type" header field

      at Test._assertHeader (node_modules/supertest/lib/test.js:237:45)
      ...

  ● middleware (fastify) › basic › should work with difference requests › should work handle requests › should return the "200" code for the "GET" request for the "noextension" url

    expected "Content-Type" header field

      at Test._assertHeader (node_modules/supertest/lib/test.js:237:45)
      ...

  ● middleware (fastify) › basic › should work with difference requests › should work handle requests › should return the "200" code for the "GET" request for the "filename-name-with-dots/mono-v6.x.x" url

    expected "Content-Type" header field

      at Test._assertHeader (node_modules/supertest/lib/test.js:237:45)
      ...

  ● middleware (fastify) › basic › should work with difference requests › should work handle requests › should return the "200" code for the "GET" request for the "noextension" url

    expected "Content-Type" header field

      at Test._assertHeader (node_modules/supertest/lib/test.js:237:45)
      ...

  ● middleware (fastify) › basic › should work with difference requests › should work handle requests › should return the "200" code for the "GET" request for the "filename-name-with-dots/mono-v6.x.x" url

    expected "Content-Type" header field

      at Test._assertHeader (node_modules/supertest/lib/test.js:237:45)
      ...

  ● middleware (fastify) › basic › should work with difference requests › should work handle requests › should return the "200" code for the "GET" request for the "noextension" url

    expected "Content-Type" header field

      at Test._assertHeader (node_modules/supertest/lib/test.js:237:45)
      ...

  ● middleware (fastify) › mimeTypes option › should set the correct value for "Content-Type" header to unknown MIME type › should return the "200" code for the "GET" request to "file.phtml"

    expected "Content-Type" header field

      at Test._assertHeader (node_modules/supertest/lib/test.js:237:45)
      ...

  ● middleware (fastify) › index option › should work with "string" value with a custom extension › should return the "200" code for the "GET" request to the public path

    expected "Content-Type" header field

      at Test._assertHeader (node_modules/supertest/lib/test.js:237:45)
      ...

  ● middleware (fastify) › index option › should work with "string" value without an extension › should return the "200" code for the "GET" request to the public path

    expected "Content-Type" header field

      at Test._assertHeader (node_modules/supertest/lib/test.js:237:45)
      ....

A worker process has failed to exit gracefully and has been force exited. This is likely caused by tests leaking due to improper teardown. Try running with --runInBand --detectOpenHandles to find leaks.
Test Suites: 1 failed, 10 passed, 11 total
Tests:       28 failed, 995 passed, 1023 total
Snapshots:   66 passed, 66 total
Time:        13.362 s, estimated 14 s
Ran all test suites.

```
It looks that fastify fixed the error with setting the content-length here : https://github.com/fastify/fastify/pull/2215
but it still doesn't support it for `head` requests.
It seems that in the new version fastify don't set `Content-Type` for unknown types at all.

- And that's what happens when I just return `res.send`.
![note send](https://user-images.githubusercontent.com/18349358/92326627-d8e41700-f053-11ea-9573-a06a04d39815.png)

Reply:
I think we should fix some tests, because in most cases `send` exists

Reply:
Yes it passes in most cases because I'm running all the tests (`getFileFromUrl.test.js`, `getPaths.test.js` ...) including `middleware.test.js` but if I test only `middleware.test.js` that actually uses the `.send` function about half of the tests pass (all express tests + 404 tests).
![only mid](https://user-images.githubusercontent.com/18349358/92382071-4b183280-f10c-11ea-9209-562338bc2ac6.png)
If I test fastify alone only `404` (144) tests pass which makes sense. 
![error 404](https://user-images.githubusercontent.com/18349358/92385498-49ea0400-f112-11ea-8d74-101fa625eece.png)

Reply:
It is very strange, because it must either completely pass or completely fall, can you look deeper? What is `res` for failed tests?